### PR TITLE
Querier: Added tenant identifier in querier http request metrics

### LINKED ISSUE #3572: Track data usage across tenant / system
> We don't actually have any way to track what % of samples or overall data is ever queried, do we? Or does something mark blocks as they're ready to know usage? 

Currently, we don't have this statistic in our system but would be nice to do the first step in this direction.

AC:
*  Allow tenant / system owner to answer question: What data is used/queried the most?

I think first is to also define the granularity of this information. I don't think the sample makes sense, but something like those most used matchers / query and the most used time range would be definitely nice to get. In some form of a histogram

### PR DESCRIPTION
**Pull request by Abhishek357**
Signed-off-by: Abhishek357 <abhisheksinghchauhan357@gmail.com>

Fixes #3572 

A promql query is executed using different endpoints of v1 API. To instrument these endpoints Thanos exposes 4 metrics (i.e.  http_request_duration_seconds, http_request_size_bytes, http_requests_total, http_response_size_bytes). This PR focuses on adding tenant identifier(tenant which is making the promql query) as a label to these metrics. These tenant identifiers are extracted from the header of requests.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

**Comment by Abhishek357** [comment]
Current status: Debugging some errors
<details>
<summary>Error while testing with quickstart.sh</summary>
<br>
panic: runtime error: index out of range [0] with length 0

goroutine 1 [level=info name=query-1 ts=2021-06-15T12:16:21.483557309Z caller=logger.go:18 msg="debug logging disabled"
running]:
github.com/thanos-io/thanos/pkg/extprom/http.(*defaultInstrumentationMiddleware).NewHandler(0xc0009e39e0, 0x1cd5a75, 0x5, 0x1fc60e0, 0xc000a018b0, 0x0, 0x0, 0x0, 0x0)
        /home/abhishek/go/src/github.com/thanos/pkg/extprom/http/instrument_server.go:99 +level=info name=query-1 ts=2021-06-15T12:16:21.4838101Z caller=logger.go:18 msg="debug logging disabled"
0x6ec
github.com/thanos-io/thanos/pkg/ui.instrf(...)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/ui.go:178
github.com/thanos-io/thanos/pkg/ui.(*Query).Register(0xc0002a6780, 0xc0009e39c0, 0x1fc19e0, 0xc0009e39e0)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/query.go:80 +0x15b
main.runQuery(0xc0009f0210, 0x1fc05c0, 0xc0009c2e70, 0xc0006f9310, 0x1fe9d20, 0xc000a00ef0, 0xc00020d368, 0x1, 0x1, 0xc000a00fa0, ...)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/query.go:563 +0x163a
main.registerQuery.func1(0xc0009f0210, 0x1fc05c0, 0xc0009c2e70, 0xc0006f9310, 0x1fe9d20, 0xc000a00ef0, 0xc00011c180, 0x1, 0x1fe9d20, 0xc000a00ef0)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/query.go:210 +0x10aa
main.main()
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/main.go:128 +0xb64
level=debug name=query-1 ts=2021-06-15T12:16:21.485836059Z caller=main.go:64 msg="maxprocs: No GOMAXPROCS change to reset%!(EXTRA []interface {}=[])"
level=info ts=2021-06-15T12:16:21.48849708Z caller=head.go:741 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2021-06-15T12:16:21.488582261Z caller=head.go:755 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=13.279µs
level=info ts=2021-06-15T12:16:21.488606255Z caller=head.go:761 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2021-06-15T12:16:21.490169285Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=7
level=info ts=2021-06-15T12:16:21.490393984Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=1 maxSegment=7
level=info ts=2021-06-15T12:16:21.490575602Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=2 maxSegment=7
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
level=info ts=2021-06-15T12:16:21.490741978Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=3 maxSegment=7
github.com/thanos-io/thanos/pkg/extprom/http.(*defaultInstrumentationMiddleware).NewHandler(0xc000b30b40, 0x1cd5a75, 0x5, 0x1fc60e0, 0xc000b2ed70, 0x0, 0x0, 0x0, 0x0)
        /home/abhishek/go/src/github.com/thanos/pkg/extprom/http/instrument_server.go:99 +0x6ec
level=info ts=2021-06-15T12:16:21.490957736Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=4 maxSegment=7
github.com/thanos-io/thanos/pkg/ui.instrf(...)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/ui.go:178
github.com/thanos-io/thanos/pkg/ui.(*Query).Register(0xc000c30180, 0xc000b30b20, 0x1fc19e0, 0xc000b30b40)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/query.go:80 +0x15b
main.runQuery(level=info ts=2021-06-15T12:16:21.491155333Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=5 maxSegment=7
0xc000116780, 0x1fc05c0, 0xc0009e11a0, 0xc000a06410, 0x1fe9d20, 0xc000b2e3a0, 0xc0005a0058, 0x1, 0x1, 0xc000b2e450, ...)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/query.go:563 +0x163a
main.registerQuery.func1(0xc000116780, 0x1fc05c0, 0xc0009e11a0, 0xc000a06410, 0x1fe9d20, level=info ts=2021-06-15T12:16:21.491361964Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=6 maxSegment=7
0xc000b2e3a0, 0xc000c10000, 0x1, 0x1fe9d20, 0xc000b2e3a0)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/query.go:210 +0x10aa
main.main()
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/main.go:128 +0xb64
level=info ts=2021-06-15T12:16:21.491829869Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=7 maxSegment=7
level=info ts=2021-06-15T12:16:21.49185388Z caller=head.go:818 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=75.634µs wal_replay_duration=3.14781ms total_replay_duration=3.284667ms
level=info ts=2021-06-15T12:16:21.493080249Z caller=rule.go:647 msg="a leftover lockfile found and removed"
level=warn ts=2021-06-15T12:16:21.493109099Z caller=rule.go:359 msg="no alertmanager configured"
level=info ts=2021-06-15T12:16:21.493848405Z caller=options.go:24 component=rules protocol=gRPC msg="disabled TLS, key and cert must be set to enable"
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/thanos-io/thanos/pkg/extprom/http.(*defaultInstrumentationMiddleware).NewHandler(0xc00011e160, 0x1cd6e3f, 0x6, 0x1fc60e0, 0xc0004e9840, 0x0, 0x0, 0x0, 0x1cd4ac8)
        /home/abhishek/go/src/github.com/thanos/pkg/extprom/http/instrument_server.go:99 +0x6ec
github.com/thanos-io/thanos/pkg/ui.instrf(...)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/ui.go:178
github.com/thanos-io/thanos/pkg/ui.(*Rule).Register(0xc000905d60, 0xc00011e0c0, 0x1fc19e0, 0xc00011e160)
        /home/abhishek/go/src/github.com/thanos/pkg/ui/rule.go:170 +0x15b
main.runRule(0xc000117788, 0x1fc05c0, 0xc0007fa4b0, 0xc000099130, 0x1fe8400, 0x2db9eb8, 0x1fc5260, 0xc000299980, 0x7ffe1e535f98, 0xd, ...)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/rule.go:568 +0x25d4
main.registerRule.func1(0xc000117788, 0x1fc05c0, 0xc0009af980, 0xc000099130, 0x1fe8400, 0x2db9eb8, 0xc00004d260, 0x0, 0x55, 0x7fef2f6be9f0)
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/rule.go:192 +0x8e8
main.main()
        /home/abhishek/go/src/github.com/thanos/cmd/thanos/main.go:128 +0xb64
all started; waiting for signal
</details>

**Comment by yashrsharma44** [comment]
Did you run the `make quickstart` in the `main` branch?

File: pkg/extprom/http/instrument_server.go [hunk 1]
```diff
@@ -20,12 +20,12 @@ import (
 // and provides necessary behaviors.
 type InstrumentationMiddleware interface {
 	// NewHandler wraps the given HTTP handler for instrumentation.
-	NewHandler(handlerName string, handler http.Handler) http.HandlerFunc
+	NewHandler(handlerName string, handler http.Handler, tenantIdentifier ...string) http.HandlerFunc
```
Comment (yashrsharma44) [inline pkg/extprom/http/instrument_server.go:23]: Let's propagate the `tenantIdentifier` using `context` variables.

**Review by yashrsharma44 — COMMENTED** [review]
Some pointers during discussion

File: pkg/api/api.go [hunk 2]
```diff
@@ -193,6 +199,14 @@ func GetRuntimeInfoFunc(logger log.Logger) RuntimeInfoFn {
 	}
 }
 
+func getTenantFromHeaders(next http.Handler) http.HandlerFunc{
+	return func(w http.ResponseWriter, r *http.Request) {
+		next.ServeHTTP(w, r.WithContext(
+			context.WithValue(r.Context(), tenantIDKey, r.Header.Get("tenantIdentifier")),
```
Comment (bwplotka) [inline pkg/api/api.go:205]: Please avoid hardcoding, this is probably important header name that we also need to document somewhere

Comment (bwplotka) [reply pkg/api/api.go:205]: Is it worth to add empty string if the header is not set?

File: pkg/api/api.go [hunk 3]
```diff
@@ -89,6 +90,11 @@ type ApiError struct {
 	Err error
 }
 
+const (
+	// tenantIDKey is the key that holds the tenant ID in a request context.
+	tenantIDKey string = "tenantID"
```
Comment (bwplotka) [inline pkg/api/api.go:95]: If we want only this package to be able to edit this, let's use `tenantIDKey = struct{}` and use only this variable

File: pkg/api/api.go (hunk 2 above)
Comment (bwplotka) [reply pkg/api/api.go:205]: We can add a flag `default-tenant-id` (Let's make sure flag name is similar to the receive)

**Review by bwplotka — COMMENTED** [review]
Some comments during Thanos Contributor Hours

* Let's mention we actually touch all components with this (since you change the global instrumentation)
* Let'\s add changelog because this is a breaking change for metrics

File: pkg/api/api.go [hunk 4]
```diff
@@ -217,11 +232,15 @@ func GetInstr(
...
 
-		return tracing.HTTPMiddleware(tracer, name, logger,
-			ins.NewHandler(name,
-				logMiddleware.HTTPMiddleware(name,
-					gziphandler.GzipHandler(
-						middleware.RequestID(hf),
+		return getTenantFromHeaders(
+			tracing.HTTPMiddleware(tracer, name, logger,
+				ins.NewHandler(name,
+					ins.InsertTenantIdentifier(
```
Comment (bwplotka) [inline pkg/api/api.go:238]: I wonder what's the plan to propagate this header/metadata for further clients (e.g StoreAPI calls)

File: pkg/extprom/http/instrument_server.go [hunk 5]
```diff
@@ -57,35 +65,57 @@ func NewInstrumentationMiddleware(reg prometheus.Registerer, buckets []float64)
...
+func getTenantIdentifier(ctx context.Context) string {
+	return ctx.Value("tenantID").(string)
+}
+
+// InsertTenantIdentifier inserts tenant identifier in metrics.
+func (ins *defaultInstrumentationMiddleware) InsertTenantIdentifier(next http.Handler) http.HandlerFunc {
+	return func (w http.ResponseWriter, r *http.Request) {
+		tenantIdentifier := getTenantIdentifier(r.Context())
+		if tenantIdentifier != "" {
+			ins.requestSize.CurryWith(prometheus.Labels{"tenant_id": tenantIdentifier})
```
Comment (bwplotka) [inline pkg/extprom/http/instrument_server.go:106]: I thought `CurryWith` is actually returning a metric which carries this label. But no one use this?

**Pull request closed without merging**